Documentation/Evolution/CharacterClasses.md

```diff
@@ -0,0 +1,354 @@
+
+## Proposed Solution
+
+We propose a treatment of character classes in Swift regular expression that is consistent with the Swift string model, configurable at the point of declaration, and comprehensible for users getting to know Swift regular expressions.
+
+### Consistency
+
+Just as the default view of a string's contents are as a collection of grapheme-cluster characters, with comparisons using canonical equivalence, regular expression matching by default will do the same. This assures that similar code produces predictable results as we add APIs that accept regular expressions and other patterns:
+
+```swift
+let str = "Cafe\u{301}"
+str.contains("é")       // true
+str.contains(/é/)       // true
+```
+
+Moreover, in accordance with the [Unicode guidelines for regular expressions](https://www.unicode.org/reports/tr18), we use Unicode property data rather than specific lists of characters whenever possible.
+
+
+### Configuration
+
+For compatibility with other languages, and in some cases to allow users to opt into higher performance, you can configure the character class matching level of regular expression at the declaration site. For example, you can opt into the Unicode scalar-based matching of other languages by using the `unicodeScalarSemantics` property on a regular expression instance.
+
+This example declares `pattern` with Unicode scalar semantics, instead of the default character semantics. `pattern` doesn't match the string above because `"e"` and `"\u{301}"` are treated as separate values.
+
+```swift
+let pattern = /^Caf.$/.unicodeScalarSemantics
+str.contains(pattern)   // false
+```
+
+Initially, we plan to allow configuration between character, Unicode scalar, and POSIX-compatible character class matching.
+
+
+### Comprehensibility
+
+For each of the character classes described below, we will base the  also add corresponding Boolean properties on both `Character` and `UnicodeScalar`. These properties will provide multiple valuable benefits:
+
+- they will be accessible from within the regular expression/pattern DSL,
+- the characters or values that each class matches can be clearly described in the symbols' documentation, and
+- the matched characters can be validated by accessing the properties directly.
+
+For example, the `\w` character class, for "word" characters, will have a matching `matchesAsWordCharacter` property.
+
+```swift
+"C".matchesAsWordCharacter  // true
+"é".matchesAsWordCharacter  // true
+"‡".matchesAsWordCharacter  // false
+```
+
+We could further extend the API to allow the redefinition of individual character classes, such as making `\d` match only the characters in the range `0` through `9`.
+
+```swift
+let fiveDigits = /\d/.customizingClass(\.matchesAsDigit, toMatch: /[0-9]/)
+```
```

Comment:
Are we talking about this? I thought we said just POSIX, grapheme, or scalar modes? Are we pitching these API?